src/layout/barGrid.ts

```diff
@@ -464,26 +515,77 @@ export function layout(seriesType: string, ecModel: GlobalModel) {
 
         lastStackCoords[stackId] = lastStackCoords[stackId] || [];
 
+        layoutInfo.groupOrder = seriesModel.get('groupOrder');
+        columnOffsetList[seriesIndex] = columnOffset;
+
         data.setLayout({
             bandWidth: columnLayoutInfo.bandWidth,
             offset: columnOffset,
             size: columnWidth
         });
-
         const valueDim = data.mapDimension(valueAxis.dim);
         const baseDim = data.mapDimension(baseAxis.dim);
         const stacked = isDimensionStacked(data, valueDim /*, baseDim*/);
         const isValueAxisH = valueAxis.isHorizontal();
 
         const valueAxisStart = getValueAxisStart(baseAxis, valueAxis, stacked);
-
+        let dataToPointList: number[];
         for (let idx = 0, len = data.count(); idx < len; idx++) {
-            const value = data.get(valueDim, idx);
+            const value = data.get(valueDim, idx) as number;
             const baseValue = data.get(baseDim, idx) as number;
-
+            if (isNaN(value) || isNaN(baseValue)) {
+                //When the value is NaN, it is still set to maintain the original code logic
+                data.setItemLayout(idx, {
+                    x: undefined,
+                    y: undefined,
+                    width: undefined,
+                    height: undefined
+                });
+                continue;
+            }
+            if (layoutDataList[baseValue] === undefined) {
+                layoutDataList[baseValue] = [];
+            }
             const sign = value >= 0 ? 'p' : 'n' as 'p' | 'n';
+            if (isValueAxisH) {
+                dataToPointList = [value, baseValue];
+            } else {
+                dataToPointList = [baseValue, value];
+            }
+            layoutDataList[baseValue].push({
+                idx: idx,
+                value: value,
+                baseValue: baseValue,
+                seriesModelData: data,
+                columnWidth: columnLayoutInfo.width,
+                coord: cartesian.dataToPoint(dataToPointList),
+                stacked: stacked,
+                stackId: stackId,
+                sign: sign,
+                valueAxisStart: valueAxisStart,
+                isValueAxisH: isValueAxisH,
+                barMinHeight: barMinHeight
+            })
+        }
+    });
+    if (layoutInfo.groupOrder !== undefined) {
+        orderLayoutData(layoutInfo);
+    }
+    zrUtil.each(layoutDataList, function (layoutDataListItem) {
+        zrUtil.each(layoutDataListItem, function (dataItem, itemIndex) {
```

Comment:
I pondered how would you unit test this long and anonymous function.

Reply:
According to the code, we need to be strong in judging the array type of 'layoutdatalistitem'. Other data nodes used are parameters that already exist in the original code. Previous unit tests for 'layout' should be able to completely cover them. However, the current 'layout' method has a large body, so split the method and simplify the code body. According to the data collection, data collation, data drawing for separation. This method may make great changes to the original code structure, but the main logic remains unchanged. I found this problem when modifying the source code. If I can, I can try splitting and refactoring.

Reply:
I would suggest to simply make anonymous function as a named function so that it can be unit tested independently. As you noted, unit test already existed to cover your changes hence it would be simply an extra mile to 'name' the function. Of course, you are free to offer fine grained unit tests before you refactor further.

Reply:
OK, I will make this anonymous function as a named function. Thanks for your help. I will submit the repair version later

Reply:
Yep, that will do.